src/base/static/libs/maps/mapboxgl-provider.js

```diff
@@ -149,52 +323,33 @@ export default (container, options) => {
       };
     },
 
-    createGeoJSONLayer: options => {
-      map.addSource(options.id, {
-        type: "geojson",
-        data: options.url,
-      });
-
-      return options.rules.map((styleRule, i) => {
-        styleRule.id = [options.id, "_", i].join("");
-        styleRule.source = options.id;
+    getSource: sourceId => {
+      return map.getSource(sourceId);
+    },
 
-        return styleRule;
+    createGeoJSONLayer: ({ id, source, rules }) => {
+      map.addSource(id, {
+        type: "geojson",
+        data: source,
       });
-    },
 
-    addLayer: layerConfig => {
-      map.addLayer(layerConfig);
-    },
+      rules = rules.map(rule => ({
+        baseLayerId: id,
+        source: id,
+        ...rule,
+      }));
 
-    addVectorLayerGroup: layerStyles => {
-      layerStyles.forEach(layerStyle => {
-        map.addLayer(layerStyle);
-      });
+      return rules
+        .map(configRuleToSymbolLayer)
+        .concat(rules.map(configRuleToLineLayer))
+        .concat(rules.map(configRuleToFillLayer));
```

Comment:
I think we should be able to infer from the layer config which symbol/line/fill layers will apply. I'm still getting familiar with our layer config, but if we have a proper representation of the layer in our config, this should be doable.

Although it might be best to punt this for now.

Reply:
The tricky thing is that a GeoJSON or place layer config can't be sure that its source data doesn't have `Point`s, `LineString`s, and/or `Polygon`s. We can't hard-code this either, because the source data can change (e.g. someone adds a polygonal featured place to a Mapseed place layer via the editor).

I think we have to inspect the GeoJSON itself at runtime to make a determination about which layer types need to be built.

Reply:
Sounds good. It sounds like we agree that we can punt it for now.

I am thinking that we can add extra complexity later once we have better abstractions.

Reply:
Perhaps we can add these notes, or a link to them, as a code comment. We can even create an issue if needed.

Reply:
I'll add some comments here-- I don't think we should forget about the need to refactor this because I think it could become a performance bottleneck.